Add AuthorizationRequired State

zmaglica opened the pull request:
Fixes #7418 

#### Changes proposed in this Pull Request

This PR adds Authentication required state that handles the situation when payment needs to be authorized (3DS). Among others, this PR introduces multiple state changes.

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

* Make sure that all tests passed
* Make sure to test with a credit card that requires 3DS (4000000000003063)
* Make sure that you are redirected to the authorization URL.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

zmaglica committed:
Updated InitalState 'processing' method

zmaglica committed:
Added get and set intent to context object

zmaglica committed:
Updated function name to match with InitialState class.

zmaglica committed:
Added Verify state that will be executed after Inital state

zmaglica committed:
Lint fixes.

RadoslavGeorgiev commented on src/Internal/Payment/PaymentContext.php:
```diff
@@ -245,4 +246,22 @@ public function set_customer_id( string $customer_id ) {
+	 *
+	 * @param WC_Payments_API_Payment_Intention $intent Instance of intent.
+	 */
+	public function set_intent( WC_Payments_API_Payment_Intention $intent ) {
```
Let's use `WC_Payments_API_Abstract_Intention` for now, to support both payment and setup intents.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/018d6303d01588e7b589c7af01d6d97303d5dee7

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -46,69 +42,45 @@ class InitialState extends AbstractPaymentState {
-
-		// If everything went well, transition to the completed state.
-		return $this->create_state( CompletedState::class );
+		return $this->create_state( VerifyState::class );
```
In this case we're not done with processing, so this should be:
```suggestion
		$next_state = $this->create_state( VerifyState::class );
		return $next_state->process();
```

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/fa1721e9d3fbd25e4c5b63fb88d5b21bf6d8e1f6

RadoslavGeorgiev commented on src/Internal/Payment/State/ProcessedState.php:
```diff
@@ -0,0 +1,60 @@
+	 * @var OrderService
+	 */
+	private $order_service;
+
```
There are some weird spacings, let's run `npm run lint:php` and remove them from the PR.

zmaglica replied:
npm run lint:php didn't linted the files, which is strange, so I manually linted each of these. Commit can be found here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/96812499ac286f9767224082ef325de6e6ff015e

RadoslavGeorgiev commented on src/Internal/Service/PaymentProcessingService.php:
```diff
@@ -66,7 +66,7 @@ public function process_payment( int $order_id, bool $manual_capture = false ) {
 		$request         = new PaymentRequest( $this->legacy_proxy );
 		$initial_state   = $this->state_factory->create_state( InitialState::class, $context );
-		$completed_state = $initial_state->process( $request );
+		$completed_state = $initial_state->start_processing( $request );
```
Seems like here we can immediately return the result of `start_processing`, avoiding a variable. This way there would be no misleading `completed_state`, which might not be completed yet.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/ce20eaf4b3aa50f5d0a85af0d1e2d988cc09df10

RadoslavGeorgiev commented on src/Internal/Service/CheckoutEncryptionService.php:
```diff
@@ -0,0 +1,44 @@
+		 * @return string
+		 */
+	public function encrypt_client_secret( string $customer_id, string $client_secret ): string {
+		if ( \WC_Payments_Features::is_client_secret_encryption_enabled() ) {
```
Please use the `LegacyProxy`'s `call_static` here.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/7296e3be6acb0360e89a3e6d5be2f16868847c0b

RadoslavGeorgiev commented on src/Internal/Service/CheckoutEncryptionService.php:
```diff
@@ -0,0 +1,44 @@
+		 */
+	public function encrypt_client_secret( string $customer_id, string $client_secret ): string {
+		if ( \WC_Payments_Features::is_client_secret_encryption_enabled() ) {
+			return openssl_encrypt(
```
Same here for `openssl_encrypt`, but with `LegacyProxy`'s `call_function()`.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/7296e3be6acb0360e89a3e6d5be2f16868847c0b

RadoslavGeorgiev commented on src/Internal/Payment/State/VerifyState.php:
```diff
@@ -0,0 +1,85 @@
+/**
+ * This state is used to create payment intent and verify next actions based on the intent state.
+ */
+class VerifyState extends AbstractPaymentState {
```
Let's make this `VerifiedState`, as verification is already done.

zmaglica replied:
Good point. Refactored here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/585514936cd51dbe088123bba8615412d3a3272b

RadoslavGeorgiev commented on src/Internal/Payment/State/VerifyState.php:
```diff
@@ -0,0 +1,85 @@
+			$intent = $this->payment_request_service->create_intent( $context );
+
+			// Something went wrong since the intent is not an instance of payment intent class. Transition to error state.
+			if ( ! $intent instanceof WC_Payments_API_Payment_Intention ) {
```
`create_intent` should either return an intent, or throw an exception. This part should not be needed.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/684b2ad638ebdad6136a352689050e87836c3dd6

RadoslavGeorgiev commented on src/Internal/Payment/State/VerifyState.php:
```diff
@@ -0,0 +1,85 @@
+		}
+
+		// All good. Proceed to processed state.
+		return $this->create_state( ProcessedState::class );
```
Now that it is processed, we can immediately try to complete it.
```suggestion
		$next_state = $this->create_state( ProcessedState::class );
		return $next_state->complete();
```

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/582352a1000f75a7d1f6c5b43c1655598557b4d1

RadoslavGeorgiev commented on src/Internal/Payment/State/ProcessedState.php:
```diff
@@ -0,0 +1,60 @@
+	 * @throws \WCPay\Exceptions\Order_Not_Found_Exception
+	 * @throws \WCPay\Internal\Payment\Exception\StateTransitionException
+	 */
+	public function complete() {
```
Let's make sure that methods like this one are defined in the base class as well.

zmaglica replied:
Adressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/427a6c36a2fbb2bf7f368a10daa4646f4120d3c4

RadoslavGeorgiev reviewed (changes requested):
Great job so far, here's my initial feedback:

RadoslavGeorgiev commented on src/Internal/Payment/State/AuthenticationRequiredState.php:
```diff
@@ -0,0 +1,75 @@
+	/**
+	 * Get response that will be sent to the client.
+	 */
+	public function get_response(): array {
```
This should probably not live in the state, as the redirect (especially 3DS) is very gateway and JS-specific. On the other hand, it is just a string, so maybe it makes sense. I'd like @marcinbot 's opinion here as well.

zmaglica replied:
Good point. Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/044cdf8890f23a366b897fef839b1c28c29bf06a

zmaglica committed:
Changed class on intent getter and setter

zmaglica committed:
Code review fix: Fixed issue with VerifyState and process method

zmaglica committed:
Code review fixes: Lint fix

zmaglica committed:
Code review fixes: Removed completed state variable in processing service

zmaglica committed:
Code review fixes: Used legacy proxy sergice to call methods in CheckoutEncryptionService

zmaglica committed:
Renamed VerifyState to VerifiedState

zmaglica committed:
Removed intent check in VerifiedState

zmaglica committed:
Code review fixes: Called complete() method on ProcessedState inside VerifiedState

zmaglica committed:
Added missing method

zmaglica committed:
Refactored AuthenticationRequiredState to return redirect url only.

zmaglica committed:
Added legacy proxy call and fixed dependencies

zmaglica commented:
> Great job so far, here's my initial feedback:

Thank you for the review. It really helped a lot. I have addressed all comments. Please let me know if there is anything to change or need additional feedback.

RadoslavGeorgiev commented on src/Internal/Payment/State/AuthenticationRequiredState.php:
```diff
@@ -0,0 +1,73 @@
+	 *
+	 * @return string
+	 */
+	public function get_redirect_url() {
```
I'm wondering whether `get_redirect_url()` is the correct name to use here, considering this method stays at all. On one hand, `get_authentication_url` would be more specific to what it does, on the other hand, we could utilize states and the fact that `get_redirect_url()` would be specific to this state.

No need to change anything here immediately, I'm just wondering.

zmaglica replied:
Make sense. I addressed it here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/e2189419b383bb031404e3c4f906b26e169caa45

RadoslavGeorgiev commented on src/Internal/Payment/State/AuthenticationRequiredState.php:
```diff
@@ -0,0 +1,73 @@
+			$context->get_amount() > 0 ? 'pi' : 'si',
+			$context->get_order_id(),
+			$this->checkout_encryption_service->encrypt_client_secret( $intent->get_customer_id(), $intent->get_client_secret() ),
+			$this->legacy_proxy->call_function( 'wp_create_nonce', 'wcpay_update_order_status_nonce' )
```
There is another thread (https://github.com/Automattic/woocommerce-payments/pull/7471/#pullrequestreview-1674442396 is the link, but GH is removing the hash/deep link) where I mentioned that some of this is too specific to live within the state. What I mean:

1. The `'#wcpay-confirm-%s:%s:%s:%s'` format is tightly tied to our JS implementation. If we wanted to change it, we'd have some very weird `if()`s deep within the payment process. There should be another layer to handle those. I'm not saying it's the gateway class, but this is probably not the right place.
2. Neither the decryption of the client secret, nor the verification of this nonce will be performed within states, but through a controller that will trigger the state machine. As it will be the job of that controller to verify the nonce, its generation should be close to the verification. Presently, that controller is the gateway class, so we should probably keep this there. Once we add 3DS post-verification to the new payment process, we'll create the proper controller, and that code should fit well there.

With that said, this is a lot of Stripe-specific code to put in the gateway. What do you think that creating the new controller(-like) class now, and putting this code there? Then in the gateway we'd be able to use the service, and we'll be able to use it later as well:

```php
public function new_process_payment( WC_Order $order ) {
	// ...
	$state = $service->process_payment( $order->get_id(), $manual_capture );

	if ( $state instanceof AuthenticationRequiredState ) {
		$redirect = $controller->generate_authentication_url( $state->get_context()->get_intent(), $order->get_id() );

		return [
			'result'   => 'success',
			'redirect' => $redirect,
		];
	}
}
```

Note: The amount wouldn't be needed because we can check whether the intent is a setup or payment intent.

zmaglica replied:
I agree with your perspective regarding the stripe related code. Relocating this code might be beneficial in the long run.

Moreover, controllers are mainly designed to receive request data, process it by using services for example, and deliver (return) results. Therefore, I suggest moving this code to a service. Since the `CompletedState` and `ErrorState` also need to return responses, it makes sense to me to move them into the same service. This way, we can manage the business logic more effectively and in a centralized way.

~~The ideal candidate, IMHO, could be `PaymentProcessingService` where we already have original state processing and parsing results at the end of the cycle could be a nice way to handle it here.~~

RadoslavGeorgiev commented on src/Internal/Service/CheckoutEncryptionService.php:
```diff
@@ -0,0 +1,57 @@
+	 *
+	 * @return string
+	 */
+	public function encrypt_client_secret( string $customer_id, string $client_secret ): string {
```
If we go with the suggestion of creating a controller for managing 3DS verification URLs, maybe we could avoid having a whole service for this, and have `encrypy_client_secret` there?

zmaglica replied:
Good catch. We don't need that service. Removed it here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/8f1ebd8f2f87833dcccb8539e183ed9e164d3218

RadoslavGeorgiev reviewed (changes requested):
I like the direction of this PR, great job adjusting it to the concept! 👏 

There are only a couple of topics I'd like to run by you in comments. In the meantime, feel free to proceed with the PR:

1. Basic checkout still works, but it would be nice to see the new `AuthenticationRequiredState` handled as well.
2. Psalm has a few errors, let's address them, and make sure that there are no left-over `@throws` in transition methods, especially within `InitialState`.
3. There are various unused `use` statements. Let's tidy up the code by removing them.

It might make sense to address as much as possible before starting to fix existing tests, and adding new ones, to avoid having to change fresh tests.

zmaglica committed:
Changed method name in AuthenticationRequiredState

zmaglica committed:
Removed CheckoutEncryptionService service and moved methods to the state

zmaglica committed:
PSALM fixes

zmaglica committed:
Added auth required response class

zmaglica committed:
Removed response class

RadoslavGeorgiev commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -32,6 +32,8 @@
 use WCPay\Duplicate_Payment_Prevention_Service;
 use WCPay\Fraud_Prevention\Fraud_Prevention_Service;
 use WCPay\Fraud_Prevention\Fraud_Risk_Tools;
+use WCPay\Internal\Payment\Response\AuthenticationRequiredResponse;
```
This line is now not used.

RadoslavGeorgiev replied:
Somehow `AuthenticationRequiredResponse` is back :D

RadoslavGeorgiev commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -826,6 +828,15 @@ public function new_process_payment( WC_Order $order ) {
 
+		if ( $state instanceof AuthenticationRequiredState ) {
+			$context = $state->get_context();
+			// Maybe we can move this inside the response class and call function like generate_response() that will something like this where we have more control of status code, ...
```
Is this line still needed?

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -46,69 +42,53 @@ class InitialState extends AbstractPaymentState {
+		/**
+		 * Next state variable.
+		 *
+		 * @var VerifiedState $next_state
```
Instead of defining this here, please make sure that all transition methods (`start_processing`, `process`, and `complete`) are present in the abstract state.

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -13,11 +13,7 @@
 use WCPay\Internal\Service\OrderService;
 use WCPay\Internal\Service\Level3Service;
 use WCPay\Internal\Service\PaymentRequestService;
```
This is an unused dependency, let's remove it.

RadoslavGeorgiev commented on src/Internal/Payment/State/VerifiedState.php:
```diff
@@ -0,0 +1,84 @@
+	 *
+	 * @var OrderService
+	 */
+	private $order_service;
```
Let's remove the dependency on the order service from this state. Everything, needed from the order service, is already loaded in `InitialState` or completed in `ProcessedState`.

RadoslavGeorgiev commented on src/Internal/Payment/State/VerifiedState.php:
```diff
@@ -0,0 +1,84 @@
+
+namespace WCPay\Internal\Payment\State;
+
+use WC_Payments_API_Abstract_Intention;
```
This line is not needed anymore.

RadoslavGeorgiev commented on src/Internal/Payment/State/VerifiedState.php:
```diff
@@ -0,0 +1,84 @@
+	 *
+	 * @return AbstractPaymentState
+	 *
+	 * @throws \WCPay\Exceptions\Order_Not_Found_Exception
```
Why do we need to refer back to the global namespace instead of adding a `use` statement?

RadoslavGeorgiev commented on src/Internal/Payment/State/VerifiedState.php:
```diff
@@ -0,0 +1,84 @@
+		}
+
+		// Intent requires authorization (3DS check).
+		if ( Payment_Intent_Status::REQUIRES_ACTION === $intent->get_status() ) {
```
`Payment_Intent_Status` is now deprecated in favor of `Intent_Status`.

RadoslavGeorgiev commented on src/Internal/Payment/State/ProcessedState.php:
```diff
@@ -0,0 +1,61 @@
+
+namespace WCPay\Internal\Payment\State;
+
+use WC_Payments_API_Payment_Intention;
```
There are a bunch of left-over `use` statements, let's remove them from here.

RadoslavGeorgiev commented on src/Internal/Payment/State/ProcessedState.php:
```diff
@@ -0,0 +1,61 @@
+ * This state is used when payment is completed on the server, and we need to update date on the plugin side.
+ */
+class ProcessedState extends AbstractPaymentState {
+
```
Let's not add empty lines in the beginning of code blocks, they only make sense as separators between blocks of  code.

RadoslavGeorgiev commented on src/Internal/Payment/State/ProcessedState.php:
```diff
@@ -0,0 +1,61 @@
+	 */
+	private $order_service;
+
+
```
There are still two empty lines here instead of one.

RadoslavGeorgiev commented on src/Internal/Payment/State/ProcessedState.php:
```diff
@@ -0,0 +1,61 @@
+	 */
+	public function complete() {
+		$context = $this->get_context();
+		// Complete processing.
```
One of those empty lines can be used here :p

RadoslavGeorgiev commented on src/Internal/Payment/State/AbstractPaymentState.php:
```diff
@@ -91,17 +91,42 @@ protected function create_state( string $state_class ) {
+	 * @return string
+	 * @throws \WCPay\Internal\Payment\Exception\StateTransitionException
+	 */
+	public function get_authentication_url() {
```
This method is not used anymore.

RadoslavGeorgiev commented on src/Internal/Payment/State/AuthenticationRequiredState.php:
```diff
@@ -0,0 +1,15 @@
+
+use WCPay\Internal\Proxy\LegacyProxy;
+/**
+ * The state, which indicates that the payment processing has been completed.
```
This comment needs to be updated for the given state.

RadoslavGeorgiev commented on src/Internal/Service/PaymentProcessingService.php:
```diff
@@ -7,7 +7,10 @@
 namespace WCPay\Internal\Service;
 
-use Exception; // Temporary exception! This service would have its own exception when more business logics are added.
+use Exception;
```
I think that after re-arranging things, nothing here throws a simple `Exception`. Could you please verify if that's the case, and remove it?

RadoslavGeorgiev replied:
Let's remove the comment as well :)

RadoslavGeorgiev commented on src/Internal/Service/PaymentProcessingService.php:
```diff
@@ -64,23 +67,62 @@ public function process_payment( int $order_id, bool $manual_capture = false ) {
+
+		return sprintf(
+			'#wcpay-confirm-%s:%s:%s:%s',
+			substr( $intent->get_id(), 0, 2 ), // intents starts with pi_ or si_ so we need only to first two letters.
```
Instead of string operations here, could we use the intent class?
```suggestion
			$intent instanceof WC_Payments_API_Setup_Intention ? 'si' : 'pi',
```

RadoslavGeorgiev reviewed (changes requested):
We're getting closer! 🙇

zmaglica committed:
Fixed broken tests

zmaglica committed:
Added tests for VerifiedState

zmaglica committed:
Added testes for ProcessedState

zmaglica committed:
Removed typo

zmaglica committed:
Removed unused comment

zmaglica committed:
Removed unused docblock

zmaglica committed:
Removed unused imports

zmaglica committed:
Removed order service dependecy from VerifiedState

zmaglica committed:
Added empty line

zmaglica committed:
Updated comment on state

zmaglica committed:
Added class check to determinate redirect url prefix.

zmaglica committed:
Fixed PSALM errors

RadoslavGeorgiev commented on src/Internal/Payment/State/VerifiedState.php:
```diff
@@ -0,0 +1,72 @@
+/**
+ * This state is used to create payment intent and verify next actions based on the intent state.
+ */
+class VerifiedState extends AbstractPaymentState {
```
@marcinbot should we follow the naming convention from pc2DNy-3e9-p2#payment-charges?

Otherwise, your comment makes sense, and the earlier we apply the new naming, the less we'd need to change.

> It would then transition to `PendingCapture` rather than `Verified`.

Actually, `VerifiedState` is when we've prepared all the data, and verified that we can continue with processing, or creating a charge/order. At this point, we don't have a charge. I see why we would want to name it a different way, but I'm not sure what would be more suitable.

RadoslavGeorgiev replied:
> But it feels like a redundant step a little bit. Most state transitions will probably require some sort of data verification as they progress through the process, so I think it might make sense to roll Verified back to Initial.

The state is not redundant. See https://github.com/Automattic/woocommerce-payments/issues/7411 and specifically "Handling existing successful intents". This is because of the way the duplicate payment prevention service works now, and needs to work, at least until we implement storage for the payment context. If there was an existing intent, detected during the verification stage, payment processing needs to be skipped, and we'd transition straight to the `ProcessedState`.

Now that I'm writing this though, I realize that we can simply do this within `InitialState`:

```php
class InitialState {
	function start_processing() {
		// prepare data...
		
		// verification
		$existing_intent = $this->duplicate_payment_prevention_service->check_for_existing_intent();
		if ( $existing_intent ) {
			$context->set_intent( $existing_intent );
			$next_state = $this->create_state( ProcessedState::class );
			$next_state->complete();
		}

		// normal processing (what we have in VerifiedState now)
		$intent = $this->payment_request_service->create_intent( $context );
		$context->set_intent( $intent );

		// Intent requires authorization (3DS check).
		if ( Payment_Intent_Status::REQUIRES_ACTION === $intent->get_status() ) {
			return $this->create_state( AuthenticationRequiredState::class );
		}

		// All good. Proceed to processed state.
		$next_state = $this->create_state( ProcessedState::class );
		return $next_state->complete();
	}
}
```

@zmaglica splitting `InitialState` into `InitialState` and `VerifiedState` was not a part of the issue definition for this PR. Why did you go ahead and implement it?

zmaglica replied:
> @zmaglica splitting `InitialState` into `InitialState` and `VerifiedState` was not a part of the issue definition for this PR. Why did you go ahead and implement it?

The reason for splitting lies in the introduction of new states. For more details about this decision, please refer to the conversation at: p1697092433168499/1696947391.111769-slack-C05997J5PCG. As you follow the conversation, you will notice that these new states have been discussed/added, and they need to be defined before we can proceed with defining the `AuthenticationRequiredState`. The only state that is not part of the authentication flow, which is part of this PR, is the `ProcessedState`, but I didn't want to leave the processing flow half-done.

I can easily edit the PR and remove the `VerifiedState`.

RadoslavGeorgiev replied:
> For more details about this decision, please refer to the conversation at: p1697092433168499/1696947391.111769-slack-C05997J5PCG. As you follow the conversation, you will notice that these new states have been discussed/added, and they need to be defined before we can proceed with defining the `AuthenticationRequiredState`.

I'm too used to having the verified state, and I mistakenly used as a part of the example in this message (p1697092433168499/1696947391.111769-slack-C05997J5PCG). However, it seems like we do not need the state for now, so let's merge it back with `InitialState`, please. @htdat can re-add it in https://github.com/Automattic/woocommerce-payments/pull/7450 if needed, but that will probably not be the case. Sorry for triggering this change, and then requesting it to be reverted.

> the only state that is not part of the authentication flow, which is part of this PR, is the ProcessedState, but I didn't want to leave the processing flow half-done.

There will be more states, as we proceed with the payment process. I understand the temptation to create them all in advance, but that's not needed yet. Let's leave the `ProcessedState`, but try to merge this PR soon, to allow @htdat to use it within the in-progress PR.

--- 

Regarding @marcinbot's request, lets rename:

- `AuthenticationRequiredState` to `PendingAuthenticationState`
- `CompletedState` to `CapturedState`
- Introduce a new (and empty for now) `PendingCaptureState` state. Based on the outcome of the intent (captured or not), `ProcessedState` should transition to either this state, or `CapturedState`.

@marcinbot how would we call the transition? `complete_processing()` instead of `complete()`?

zmaglica replied:
I have removed `VerifiedState` here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/7765bee4c4e6a445879d0019d080de8be2660c52

zmaglica committed:
Replaced dockblock excpetions with use statments

zmaglica committed:
Used better intent status constant

zmaglica committed:
Cleaned ProcessedState class

zmaglica committed:
Removed unused comment

zmaglica committed:
Added missing tests

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/InitialStateTest.php:
```diff
@@ -132,52 +105,68 @@ public function test_process( Exception $exception = null ) {
+					$this->mock_order_service->expects( $this->once() )
+						->method( 'update_order_from_successful_intent' )
+						->with( $this->mock_context->get_order_id(), $this->mock_context->get_intent(), $this->mock_context );
+					$processed_state = new ProcessedState( $this->mock_state_factory, $this->mock_order_service );
```
This is not testing the `InitialState`, it's testing the whole process, which is not needed, and over-complicates this test. Let's just return a mock state from the first `create_state` call, assert that the right method is called, and be done with it. Otherwise, once we add all the needed logic to the next state, this test will grow out of control.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/776c96fa45b2f0da36dd2483b2d60a645bf04abf

zmaglica committed:
Removed VerifiedState

zmaglica committed:
Simplifed test for InitialState

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -56,10 +56,10 @@ class InitialState extends AbstractPaymentState {
-	 * @param OrderService                 $order_service           Service for order-related actions.
-	 * @param WC_Payments_Customer_Service $customer_service        Service for managing remote customers.
-	 * @param Level3Service                $level3_service          Service for Level3 Data.
+	 * @param StateFactory                 $state_factory Factory for payment states.
+	 * @param OrderService                 $order_service Service for order-related actions.
+	 * @param WC_Payments_Customer_Service $customer_service Service for managing remote customers.
+	 * @param Level3Service                $level3_service Service for Level3 Data.
```
Let's restore the right spacing here.

RadoslavGeorgiev replied:
You fixed it for `start_processing`, but it should be consistent for all files within the PR (and `src` for that matter).

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -78,37 +78,42 @@ public function __construct(
 		// Populate further details from the order.
 		$this->populate_context_from_order();
 
+		$context = $this->get_context();
```
Now that the context is only used for creating the intent, let's move this line within the `try {}` block.

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -78,37 +78,42 @@ public function __construct(
 
-		// If everything went well, transition to the completed state.
-		return $this->create_state( CompletedState::class );
+		return $next_state->complete();
```
Based on conversations so far, wdyt about calling this `complete_processing`?

RadoslavGeorgiev reviewed (changes requested):
I still need to check tests, but the rest of the changes look pretty well. I only have a couple of nitpicks.

 know we are also discussing renaming states and etc., but that should be done through another issue & PR. This one is already pretty big.

zmaglica committed:
Code review fixes

zmaglica committed:
Merged with latest dev

RadoslavGeorgiev commented on src/Internal/DependencyManagement/ServiceProvider/PaymentsServiceProvider.php:
```diff
@@ -14,11 +14,14 @@
+use WCPay\Internal\Payment\State\ProcessedState;
 use WCPay\Internal\Payment\State\StateFactory;
 use WCPay\Internal\Payment\State\SystemErrorState;
+use WCPay\Internal\Payment\State\VerifiedState;
```
Small left-over here :)

zmaglica replied:
Fixed (https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d)

RadoslavGeorgiev commented on src/Internal/DependencyManagement/ServiceProvider/PaymentsServiceProvider.php:
```diff
@@ -41,6 +44,7 @@ class PaymentsServiceProvider extends AbstractServiceProvider {
 		Router::class,
 		StateFactory::class,
 		InitialState::class,
+		ProcessedState::class,
```
Just like this declaration, we need to declare that the provider provides `AuthenticationRequiredState`. Right now, everything works, because the `register` method is called for another state. At some point in the future, if someone requests the `AuthenticationRequiredState` directly, this could lead to a fatal error.

RadoslavGeorgiev commented on src/Internal/Payment/State/AbstractPaymentState.php:
```diff
@@ -91,17 +91,43 @@ protected function create_state( string $state_class ) {
+	 * @throws \WCPay\Exceptions\Order_Not_Found_Exception
+	 * @throws \WCPay\Internal\Payment\Exception\StateTransitionException
+	 */
+	public function process() {
```
This method is not used anywhere, let's only leave `start_processing` here.

zmaglica replied:
Removed (https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d)

RadoslavGeorgiev commented on src/Internal/Payment/State/AbstractPaymentState.php:
```diff
@@ -91,17 +91,43 @@ protected function create_state( string $state_class ) {
+	 *
+	 * @return AbstractPaymentState
+	 * @throws \WCPay\Exceptions\Order_Not_Found_Exception
+	 * @throws \WCPay\Internal\Payment\Exception\StateTransitionException
```
Let's be consistent, and add `use` statements for those, only listing exceptions like `Order_Not_Found_Exception` or `StateTransitionException`, keeping in mind that those are already "imported" into the file. This applies to the whole PR.

zmaglica replied:
Used `use` statements as you suggested (https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d)

RadoslavGeorgiev commented on src/Internal/Payment/State/InitialState.php:
```diff
@@ -120,6 +124,7 @@ public function process( PaymentRequest $request ) {
 	 * on all needed parameters being in place.
 	 *
 	 * @param PaymentRequest $request The request to use.
+	 *
```
More RadoCS: We don't need an empty line between parameters and exceptions. Same for the whole PR (if there are other instances).

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/97345dbc66b1cc740585d7850147f228ccdad2eb 

But I will also check the whole PR and fix if I detect some.

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/AbstractPaymentStateTest.php:
```diff
@@ -80,9 +80,22 @@ public function test_create_state_uses_the_factory() {
+	public function state_defined_methods_provider() {
+		return [
+			[ 'start_processing' ],
+			[ 'process' ],
```
This one should go away, we only have `start_processing` and `complete_processing` right now.

With that in mind, let's just have 2 separate tests for the two methods. We shouldn't assume every transition method requires either the request as a parameter, or nothing.

zmaglica replied:
This is addressed.

RadoslavGeorgiev replied:
@zmaglica I see that the methods are updated, but the bigger point of the comment was not to use a data provider, instead having two separate tests for the two methods.

zmaglica replied:
Good point. I missed that. Corrected here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/2c165178f28b4e6a39827c60fd9e55320d218bff

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/PaymentProcessingServiceTest.php:
```diff
@@ -107,11 +109,90 @@ public function test_process_payment_happy_path_without_mock_builder() {
 	}
+
+	public function test_get_authentication_redirect_url_will_return_url_from_payment_intent() {
+		$sut = new PaymentProcessingService( $this->mock_state_factory, $this->mock_legacy_proxy );
```
`$sut` is defined explicitly in the previous test because it tests `create_payment_context` (that's mocked normally). For this and all following tests, you can use `$this->sut` directly.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/b35ba641977dfd7e93c02b85d336adae06f2dcf7

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/PaymentProcessingServiceTest.php:
```diff
@@ -107,11 +109,90 @@ public function test_process_payment_happy_path_without_mock_builder() {
+				[ 'wp_create_nonce', 'wcpay_update_order_status_nonce' ]
+			)
+			->willReturnOnConsecutiveCalls( $secret, $nonce );
+		$result = $sut->get_authentication_redirect_url( $intent, $order );
```
RadoCS: In this, and the next test, let's add an empty line before `$result = ...`.

zmaglica replied:
FIxed. https://github.com/Automattic/woocommerce-payments/pull/7471/commits/97345dbc66b1cc740585d7850147f228ccdad2eb

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/PaymentProcessingServiceTest.php:
```diff
@@ -107,11 +109,90 @@ public function test_process_payment_happy_path_without_mock_builder() {
+		$result = $sut->get_authentication_redirect_url( $intent, $order );
+		$this->assertSame( "#wcpay-confirm-si:$order:" . $intent->get_client_secret() . ":$nonce", $result );
+	}
+
```
RadoCS: Too many empty lines.

zmaglica replied:
Fixed in this commit: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/97345dbc66b1cc740585d7850147f228ccdad2eb

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/PaymentProcessingServiceTest.php:
```diff
@@ -107,11 +109,90 @@ public function test_process_payment_happy_path_without_mock_builder() {
+			)
+			->willReturnOnConsecutiveCalls( $secret, $nonce );
+		$result = $sut->get_authentication_redirect_url( $intent, $order );
+		$this->assertSame( "#wcpay-confirm-pi:$order:$secret:$nonce", $result );
```
Can we add a data provider here, and test with both `pi` and `si` through it? This way `test_get_authentication_redirect_url_will_return_url_with_secret_and_si_prefix` (is the method name correct there btw?) can focus on what happens when encryption is disabled.

zmaglica replied:
I used dataprovider as you suggested and changed test names. Now the test for non encrypted secret key is standalone and PI/SI prefix tests are handled by provider.

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/InitialStateTest.php:
```diff
@@ -7,15 +7,16 @@
+use WCPay\Constants\Intent_Status;
+use WCPay\Internal\Payment\State\AuthenticationRequiredState;
+use WCPay\Internal\Payment\State\ProcessedState;
+use WCPay\Internal\Payment\State\VerifiedState;
```
Unused `use` statement.

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/InitialStateTest.php:
```diff
@@ -57,12 +58,12 @@ class InitialStateTest extends WCPAY_UnitTestCase {
 	 * @var Level3Service|MockObject
 	 */
-	private $mock_level3_service;
+	private $mock_payment_request_service;
```
Not sure why this got switched with the next property, but the `@var` definition is now incorrect.

RadoslavGeorgiev replied:
This still needs to be fixed.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/97345dbc66b1cc740585d7850147f228ccdad2eb

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/InitialStateTest.php:
```diff
@@ -254,4 +253,20 @@ public function test_populate_context_from_order() {
 		PHPUnit_Utils::call_method( $this->sut, 'populate_context_from_order', [] );
 	}
+
+	private function mock_initial_state() {
```
Re-defining the service under tests like this:

1. Makes it a bit more hidden, and not so obvious what's being done, as it's in the end of the test.
2. Causes a bunch of errors when editing:

<img width="489" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/5311119/03b347d0-6584-4c3a-985f-fb3e982bb98b">

As there are a multiple methods that use the pure class as `sut`, and multiple others that need the mocked one, let's just define both of them in the `setUp` method, something like this:

```php
/**
 * Service under test.
 *
 * @var InitialState
 */
private $sut;

/**
 * Service under test with mocked protected methods.
 *
 * @var InitialState|MockObject
 */
private $mocked_sut;

/**
 * Set up the test.
 */
protected function setUp(): void {
	parent::setUp();

	// set up mocks...

	$this->sut = new InitialState(
		$this->mock_state_factory,
		$this->mock_order_service,
		$this->mock_customer_service,
		$this->mock_level3_service,
		$this->mock_payment_request_service
	);
	$this->sut->set_context( $this->mock_context );

	$this->mocked_sut = $this->getMockBuilder( InitialState::class )
		->onlyMethods( [ 'populate_context_from_request', 'populate_context_from_order' ] )
		->setConstructorArgs(
			[
				$this->mock_state_factory,
				$this->mock_order_service,
				$this->mock_customer_service,
				$this->mock_level3_service,
				$this->mock_payment_request_service,
			]
		)
		->getMock();
	$this->mocked_sut->set_context( $this->mock_context );
}
```

Then tests can decide which one to use.

zmaglica replied:
I have updated the test with the example you provided (https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d)

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/InitialStateTest.php:
```diff
@@ -92,92 +93,90 @@ protected function setUp(): void {
-			)
-			->getMock();
-		$this->sut->set_context( $this->mock_context );
+		$this->mock_initial_state();
```
The comment above this line is useless now, let's remove it.

zmaglica replied:
The comment is removed (ref: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d)

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/ProcessedStateTest.php:
```diff
@@ -0,0 +1,84 @@
+use WCPAY_UnitTestCase;
+
+/**
+ * Tests for the initial payment state.
```
Outdated comment.

zmaglica replied:
Comment is updated now: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/ProcessedStateTest.php:
```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * Class ProcessedStateTestTest
```
`TestTest` :)

zmaglica replied:
Fixed :) https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/ProcessedStateTest.php:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	public function test_complete_processing_will_transition_to_completed_state() {
+
```
RadoCS: Let's remove this empty line.

zmaglica replied:
FIxed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d

RadoslavGeorgiev reviewed (changes requested):
Looks pretty well, only a few nitpicks left :)

Also, this would be a good time to update the PR description, and mark it as ready for review (not draft).

htdat commented on src/Internal/Payment/PaymentContext.php:
```diff
@@ -245,4 +246,22 @@ public function set_customer_id( string $customer_id ) {
 	}
+
+	/**
+	 * Stores the payment intent.
```
Nitpick: 

```suggestion
	 * Stores the payment intent object.
```

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/48da0fc92461c2e53787cf45c2e79dcf0fa5bf8d

zmaglica committed:
Lint fixes

zmaglica committed:
Added changelog

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7471` or branch name `rpp/7414-authentication-required-state` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=rpp/7414-authentication-required-state) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 518dff16e722f23d30bb7cc60995d93e75c24f96
- Build time: 2023-10-23 12:38:38 UTC

_Note: the build is updated when a new commit is pushed to this PR._

zmaglica committed:
Fixed spacing in initial state

zmaglica committed:
Code review fixes

zmaglica committed:
Fixed psalm errors

zmaglica committed:
Merge branch 'develop' into rpp/7414-authentication-required-state

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/InitialStateTest.php:
```diff
@@ -38,6 +38,12 @@ class InitialStateTest extends WCPAY_UnitTestCase {
 	 * @var InitialState
 	 */
 	private $sut;
+	/**
```
RadoCS: Let's add a line before this comment.

zmaglica replied:
https://github.com/Automattic/woocommerce-payments/pull/7471/commits/97345dbc66b1cc740585d7850147f228ccdad2eb

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/InitialStateTest.php:
```diff
@@ -38,6 +38,12 @@ class InitialStateTest extends WCPAY_UnitTestCase {
+	/**
+	 * Service under test.
+	 *
+	 * @var StateFactory|InitialState
```
```suggestion
	 * @var MockObject|InitialState
```

Just curious, doesn't PHPStorm highlight errors because of this? I'm noticing that @jessy-p's tests have something similar.

zmaglica replied:
It should, but perhaps it doesn't check tests. I will check that.

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/ProcessedStateTest.php:
```diff
@@ -0,0 +1,83 @@
+
+		$this->mock_state_factory = $this->createMock( StateFactory::class );
+		$this->mock_order_service = $this->createMock( OrderService::class );
+		$this->mock_context       = $this->createMock( PaymentContext::class );
```
The `$mock_context` property is not defined, let's define it.

zmaglica replied:
Addressed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/97345dbc66b1cc740585d7850147f228ccdad2eb

RadoslavGeorgiev reviewed (changes requested):
Only a couple of small (new) things I noticed. There are some comments from previous reviews that are not addressed, probably because they were collapsed:

![Uploading image.png…]()

zmaglica committed:
Code review fixes

zmaglica committed:
Used mocked version of state for tests

htdat commented on src/Internal/Payment/State/AuthenticationRequiredState.php:
```diff
@@ -0,0 +1,14 @@
+/**
+ * The state, which indicates that the payment requires authentication (3DS).
+ */
+class AuthenticationRequiredState extends AbstractPaymentState {
```
Based on https://github.com/Automattic/woocommerce-payments/pull/7471#discussion_r1361620696: 

```suggestion
class PendingAuthenticationState extends AbstractPaymentState {
```

htdat commented on src/Internal/Payment/State/ProcessedState.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * This state is used when payment is completed on the server, and we need to update date on the plugin side.
+ */
+class ProcessedState extends AbstractPaymentState {
```
Based on https://github.com/Automattic/woocommerce-payments/pull/7471#discussion_r1361620696: 


```suggestion
class CapturedState extends AbstractPaymentState {
```

RadoslavGeorgiev replied:
Sorry for mistakenly giving my 👍 here, but this is not right.

Because of the patch, I am still to share the state graph so we can discuss it, so let's undo this change for now. `CapturedState` should come after `ProcessedState`, it is not the same.

It would be:
1. `InitialState`
2. `PendingAuthenticationState`
3. `ProcessedState`
4. `CapturedState` or `PendingCaptureState`

The point is not to end up in a `CompletedState`, as a payment is never "complete".

Let's undo this change. We'll finalize state names in a different issue/PR.

zmaglica replied:
I have reverted a commit that changed states name.

RadoslavGeorgiev commented on tests/unit/src/Internal/Payment/State/InitialStateTest.php:
```diff
@@ -90,41 +97,8 @@ protected function setUp(): void {
-		 * @var MockObject|InitialState
-		 */
-		$this->sut = $this->getMockBuilder( InitialState::class )
+		$this->mocked_sut = $this->getMockBuilder( InitialState::class )
```
We still need the comment with the `@var` declaration here, at least for VS Code: 
<img width="779" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/5311119/1c443e84-9def-4bad-b8fc-98ddbb19004f">

Even though the private property is defined as `InitialState|MockObject`, the definition here makes the editor thing it's just a `MockObject`

You can simply restore the comment that was there:
```suggestion
		/**
		 * This test works with the root `process` method, which calls a few
		 * internal methods. We want to mock them for the purpose of this test.
		 *
		 * @var MockObject|InitialState
		 */
		$this->mocked_sut = $this->getMockBuilder( InitialState::class )
```

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/PaymentProcessingServiceTest.php:
```diff
@@ -92,8 +96,6 @@ public function test_process_payment_happy_path() {
 	 * Test the basic happy path of processing a payment.
 	 */
 	public function test_process_payment_happy_path_without_mock_builder() {
-		$sut = new PaymentProcessingService( $this->mock_state_factory, $this->mock_legacy_proxy );
```
This is the one method that should indeed use the `PaymentProcessingService` directly. Now coverage has gone down for the file, because `create_payment_context` is not covered with tests.

zmaglica replied:
Addressed here. https://github.com/Automattic/woocommerce-payments/pull/7471/commits/2c165178f28b4e6a39827c60fd9e55320d218bff

zmaglica committed:
Added docblock for mocked sut variable

Co-authored-by: Radoslav Georgiev <rageorgiev@gmail.com>

zmaglica committed:
Updated tests

zmaglica committed:
Changed state names

RadoslavGeorgiev commented on tests/unit/src/Internal/Service/PaymentProcessingServiceTest.php:
```diff
@@ -68,23 +72,19 @@ public function test_process_payment_happy_path() {
 		$mock_initial_state   = $this->createMock( InitialState::class );
 		$mock_completed_state = $this->createMock( CompletedState::class );
```
The mock context (line 71) is not used, let's remove that line.

zmaglica replied:
Good catch. Fixed here: https://github.com/Automattic/woocommerce-payments/pull/7471/commits/518dff16e722f23d30bb7cc60995d93e75c24f96

zmaglica committed:
Merge branch 'develop' into rpp/7414-authentication-required-state

zmaglica committed:
Revert "Changed state names"

This reverts commit f6abcee907c010cf8d2c5a5aafab289f1b864943.

zmaglica committed:
Removed unusued variable from test

RadoslavGeorgiev reviewed (approved):
LGTM! :shipit: 

Thank you for bearing with me through all the nitpicks :)

zmaglica commented:
> LGTM! :shipit:
> 
> Thank you for bearing with me through all the nitpicks :)

I love nitpicks 😃  Thank you for the extended review 🙇
